enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,381 @@
+
+## Proposal
+
+* Use modularization, so MicroShift no longer needs to carry plug-in specific
+  code and plug-ins are self-contained.
+
+* Make plug-in authors use a similar mechanism to deploy necessary manifests as end-users when they want
+ to extend the core (--> a generalised/evolved form of the kustomizer running a plugin from MicroShift
+ to generate the kustomize yaml).
+
+* Make plug-in authors responsible for the OS-level aspects, in particular how to deal with systemd.
+
+* Use Linux-native software package management mechanisms (here: RPM) to modularise plug-ins, allowing users to express which plug-in they want by explicitly installing / not installing an RPM.
+
+### Proposal details
+
+To allow modularity of the described core components the components are split across
+multiple MicroShift subpackages, supporting at least the following use cases:
+
+| Use case 	                                        | Packages installed in the RHEL4Edge image 	| Notes 	                            |
+|--------------------------------------------------	|-----------------------------------------  	|-------	                            |
+| The whole stack of components is needed         	| `microshift`, `microshift-networking-ovnk`, `microshift-storage-topolvm` | Being explicit about the components avoids unexpected changes to customers if the default components change; i.e. they would get an error if their component of choice got deprecated. |
+| Only advanced networking is needed                | `microshift`, `microshift-networking-ovnk`  |                                    	|
+| The bare minimum set of components is necessary   | `microshift`                             |                                   	|
+| Only persistent volumes are necessary             | `microshift`, `microshift-storage-topolvm` |                                     |
+
+
+The packages for networking and storage would be different in upstrean, containing
+references to unrestricted (OKD) container images, removing the need for a pull secret.
+
+Reuse the kustomize loader so system and user components use the mechanism,
+extending its behaviour so binary plugins can be used in this content.
+
+
+`/usr/lib/microshift/components.d/004-multus/plugin`
+`/usr/lib/microshift/components.d/005-networking-ovnk/plugin`
+`/usr/lib/microshift/components.d/010-storage-topolvm/plugin`
+
+  MicroShift component manager would scan the `/usr/lib/microshift/components.d/`
+  directory executing the component plugins in alphabetical order.
+
+  `plugin` is an executable (in most cases a bash script will be the simplest
+  solution), using go binaries would be an option when complex structures
+  or connectivity back to the MicroShift API is required.
+
+  Today for the CNI/CSI components there is no wait or readiness conditions,
+  if this needed to be implemented in the future it could be implemented in
+  the `plugin`.
+
+The following changes would be introduced to packaging:
```

Comment:
If I'm not mistaken, we need a bullet that describes how we intend to handle the release-info JSON in a way users can still obtain the list of all container images, including those by the plug-ins.